vm_insnhelper.c

```diff
@@ -6204,18 +6204,38 @@ VALUE rb_setup_fake_ary(struct RArray *fake_ary, const VALUE *list, long len, bo
 VALUE rb_ec_pack_ary(rb_execution_context_t *ec, VALUE ary, VALUE fmt, VALUE buffer);
 
 VALUE
-rb_vm_opt_newarray_pack(rb_execution_context_t *ec, rb_num_t num, const VALUE *ptr, VALUE fmt)
+rb_vm_opt_newarray_pack_buffer(rb_execution_context_t *ec, rb_num_t num, const VALUE *ptr, VALUE fmt, VALUE buffer)
 {
     if (BASIC_OP_UNREDEFINED_P(BOP_PACK, ARRAY_REDEFINED_OP_FLAG)) {
         struct RArray fake_ary;
         VALUE ary = rb_setup_fake_ary(&fake_ary, ptr, num, true);
-        return rb_ec_pack_ary(ec, ary, fmt, Qnil);
+        return rb_ec_pack_ary(ec, ary, fmt, buffer);
     }
     else {
-        return rb_vm_call_with_refinements(ec, rb_ary_new4(num, ptr), idPack, 1, &fmt, RB_PASS_CALLED_KEYWORDS);
+        // The opt_newarray_send insn drops the keyword args so we need to rebuild them.
+        // Setup an array with room for keyword hash.
+        VALUE args[2];
+        args[0] = fmt;
+        int kw_splat = RB_NO_KEYWORDS;
+        int argc = 1;
+
+        if (!NIL_P(buffer)) {
+            args[1] = rb_hash_new_with_size(1);
+            rb_hash_aset(args[1], ID2SYM(idBuffer), buffer);
+            kw_splat = RB_PASS_KEYWORDS;
+            argc++;
+        }
+
+        return rb_vm_call_with_refinements(ec, rb_ary_new4(num, ptr), idPack, argc, args, kw_splat);
     }
 }
 
+VALUE
+rb_vm_opt_newarray_pack(rb_execution_context_t *ec, rb_num_t num, const VALUE *ptr, VALUE fmt)
```

Comment:
```suggestion
static VALUE
vm_opt_newarray_pack(rb_execution_context_t *ec, rb_num_t num, const VALUE *ptr, VALUE fmt)
```

We export a symbol and use an `rb_` prefix to avoid collision when other object files like YJIT use it. Now that YJIT stopped using `rb_vm_opt_newarray_pack`, it seems confusing that it's still exported.

Nobu added a `rb_vm_opt_newarray_pack` bindgen to RJIT but he didn't seem to use it, so you can just remove  [this line](https://github.com/ruby/ruby/blob/86a762ce56cd44db69ccff2e02587ed748a3ad04/tool/rjit/bindgen.rb#L544) and run `make rjit-bindgen` again (or use its output on the CI). If you're unsure, I can handle this myself too.

Reply:
I was going to add the function references to the rjit code like #8734 did until I realized rjit wasn't using the insn:
https://github.com/ruby/ruby/blob/e798f45cc2417ae778e4790e3159654c90ba91fe/lib/ruby_vm/rjit/insn_compiler.rb#L1526-L1537

When I tried `make rjit-bindgen` I was getting this error:
```
bindgen.rb:249:in `block (2 levels) in generate_node': non-leading bit fields are not supported. consider including 'is_from_method' in skip_fields. (RuntimeError)
```
so I thought I'd leave that for another PR.

Reply:
I changed to use a static `vm_opt_newarray_pack_buffer` with the `rb_*` ones just being non-static wrappers for consistency with the other `vm_opt_newarray_*` functions.